### PR TITLE
search: match repos concurrently

### PR DESCRIPTION
**Pull request by stefanhengl**
This PR adds concurrency to repo matching. According to tracing data, repo search is a bottleneck for global queries. In a local benchmark, the implementation showed a speed-up of 4x. We have to see how this translates to performance gains on Cloud.

For global queries, for which we don't have to care about revisions, we might offload the repo matching to Zoekt, but this will be tackled in a different PR.

I also implemented a variant of this PR that uses channels instead of mutexes. I was wondering whether the diminishing gains in performance (>5 workers) were due to goroutines busy-waiting on mutexes. However, the alternative turned out to perform significantly worse.

### CODE REVIEW

File: cmd/frontend/graphqlbackend/search_repositories.go [hunk 1]
```diff
@@ -102,6 +96,68 @@ func searchRepositories(ctx context.Context, args *search.TextParameters, limit
...
+		muCommon = sync.Mutex{}
+		common   = &searchResultsCommon{}
+
+		muRepos = sync.Mutex{}
+		repos   []*search.RepositoryRevisions
+
+		muCursor = sync.Mutex{}
+		cursor   = 0
+	)
+	common.repos = make([]*types.Repo, len(resolved))
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search_repositories.go:110]: In the common case this slice is empty?

**Review by keegancsmith — APPROVED** [review]
Would performance be better if we just divide the slice into 5 and avoid the cursor mutex? Could even just append all 5 found results at the end, and avoid all mutexes

File: cmd/frontend/graphqlbackend/search_repositories.go (hunk 1 above)
Comment (stefanhengl) [reply cmd/frontend/graphqlbackend/search_repositories.go:110]: No. `common.repos` will always have the same length as `resolved`. It is a projection from `resolved.Repos` and will always contain all repos even if nothing matches.

**Comment by stefanhengl** [comment]
> Would performance be better if we just divide the slice into 5 and avoid the cursor mutex? Could even just append all 5 found results at the end, and avoid all mutexes

Apparently not. The performance is statistically the same at least for higher levels of concurrency. Interestingly, the performance without the mutex seems to have a higher variance.

old: with cursor mutex
new: without cursor mutex

```
name                   old time/op  new time/op  delta
SearchRepositories      135ms ± 5%   142ms ± 4%  +5.81%  (p=0.000 n=10+9)
SearchRepositories-2   75.2ms ± 3%  77.4ms ± 2%  +2.93%  (p=0.007 n=10+10)
SearchRepositories-3   53.1ms ± 1%  55.1ms ± 4%  +3.78%  (p=0.000 n=8+9)
SearchRepositories-4   43.8ms ±11%  47.3ms ±12%  +7.92%  (p=0.016 n=10+11)
SearchRepositories-5   38.8ms ± 8%  41.3ms ±14%    ~     (p=0.061 n=10+11)
SearchRepositories-6   39.8ms ± 7%  40.8ms ±22%    ~     (p=0.863 n=10+11)
SearchRepositories-7   40.0ms ±13%  41.3ms ±10%    ~     (p=0.349 n=10+11)
SearchRepositories-8   39.5ms ± 6%  39.0ms ±17%    ~     (p=0.756 n=10+11)
SearchRepositories-9   38.1ms ± 8%  38.2ms ±11%    ~     (p=0.796 n=10+10)
SearchRepositories-10  38.9ms ± 8%  37.6ms ± 2%    ~     (p=0.133 n=10+9)
```

**Comment by keegancsmith** [comment]
LGTM. Can you share the diff for removing the mutex cursor? I'd be interested to micro optimize since I find that surprising :P

**Comment by stefanhengl** [comment]
```
diff --git a/cmd/frontend/graphqlbackend/search_repositories.go b/cmd/frontend/graphqlbackend/search_repositories.go
index a7bb985905..e8b685c9b4 100644
--- a/cmd/frontend/graphqlbackend/search_repositories.go
+++ b/cmd/frontend/graphqlbackend/search_repositories.go
@@ -103,19 +103,14 @@ func matchRepos(pattern *regexp.Regexp, resolved []*search.RepositoryRevisions)
 
 		muRepos = sync.Mutex{}
 		repos   []*search.RepositoryRevisions
-
-		muCursor = sync.Mutex{}
-		cursor   = 0
 	)
 	common.repos = make([]*types.Repo, len(resolved))
 	wg := sync.WaitGroup{}
-	worker := func() {
+	worker := func(start int, step int) {
 		defer wg.Done()
+		j := start - step
 		for {
-			muCursor.Lock()
-			j := cursor
-			cursor += 1
-			muCursor.Unlock()
+			j += step
 			if j >= len(resolved) {
 				return
 			}
@@ -152,7 +147,7 @@ func matchRepos(pattern *regexp.Regexp, resolved []*search.RepositoryRevisions)
 	*/
 	for w := 0; w < 5; w++ { // for benchmarking, replace 5 with runtime.GOMAXPROCS(0)
 		wg.Add(1)
-		go worker()
+		go worker(w, 5)
 	}
 	wg.Wait()
 	return common, repos
```

**Comment by keegancsmith** [comment]
I couldn't help myself and implemented my idea without the use of mutexes (and a small channel):

```
name                  old time/op  new time/op  delta
SearchRepositories-5  48.6ms ± 8%  43.6ms ±21%   ~     (p=0.151 n=5+5)
```

old is the current branch, new is with the patch below applied.

```diff
diff --git a/cmd/frontend/graphqlbackend/search_repositories.go b/cmd/frontend/graphqlbackend/search_repositories.go
index a7bb985905..40401c2d7e 100644
--- a/cmd/frontend/graphqlbackend/search_repositories.go
+++ b/cmd/frontend/graphqlbackend/search_repositories.go
@@ -4,7 +4,7 @@ import (
 	"context"
 	"math"
 	"regexp"
-	"sync"
+	"runtime"
 
 	"github.com/sourcegraph/sourcegraph/internal/api"
 
@@ -97,38 +97,6 @@ func searchRepositories(ctx context.Context, args *search.TextParameters, limit
 }
 
 func matchRepos(pattern *regexp.Regexp, resolved []*search.RepositoryRevisions) (*searchResultsCommon, []*search.RepositoryRevisions) {
-	var (
-		muCommon = sync.Mutex{}
-		common   = &searchResultsCommon{}
-
-		muRepos = sync.Mutex{}
-		repos   []*search.RepositoryRevisions
-
-		muCursor = sync.Mutex{}
-		cursor   = 0
-	)
-	common.repos = make([]*types.Repo, len(resolved))
-	wg := sync.WaitGroup{}
-	worker := func() {
-		defer wg.Done()
-		for {
-			muCursor.Lock()
-			j := cursor
-			cursor += 1
-			muCursor.Unlock()
-			if j >= len(resolved) {
-				return
-			}
-			muCommon.Lock()
-			common.repos[j] = resolved[j].Repo
-			muCommon.Unlock()
-			if pattern.MatchString(string(resolved[j].Repo.Name)) {
-				muRepos.Lock()
-				repos = append(repos, resolved[j])
-				muRepos.Unlock()
-			}
-		}
-	}
 	/*
 		Local benchmarks showed diminishing returns for higher levels of concurrency.
 		5 workers seems to be a good trade-off for now. We might want to revisit this
@@ -150,12 +118,45 @@ func matchRepos(pattern *regexp.Regexp, resolved []*search.RepositoryRevisions)
 		   BenchmarkSearchRepositories-9     	      30	  38050861 ns/op
 		   BenchmarkSearchRepositories-10    	      27	  38638799 ns/op
 	*/
-	for w := 0; w < 5; w++ { // for benchmarking, replace 5 with runtime.GOMAXPROCS(0)
-		wg.Add(1)
-		go worker()
+	step := len(resolved) / runtime.GOMAXPROCS(0) // for benchmarking, replace 5 with runtime.GOMAXPROCS(0)
+	if step == 0 {
+		step = len(resolved)
+	} else {
+		step += 1
+	}
+
+	results := make(chan []*search.RepositoryRevisions)
+	workers := 0
+	offset := 0
+	for offset < len(resolved) {
+		next := offset + step
+		if next > len(resolved) {
+			next = len(resolved)
+		}
+		workers++
+		go func(repos []*search.RepositoryRevisions) {
+			var matched []*search.RepositoryRevisions
+			for _, r := range repos {
+				if pattern.MatchString(string(r.Repo.Name)) {
+					matched = append(matched, r)
+				}
+			}
+			results <- matched
+		}(resolved[offset:next])
+		offset = next
+	}
+
+	repos := make([]*types.Repo, len(resolved))
+	for i := range resolved {
+		repos[i] = resolved[i].Repo
 	}
-	wg.Wait()
-	return common, repos
+
+	var matched []*search.RepositoryRevisions
+	for w := 0; w < workers; w++ {
+		matched = append(matched, <-results...)
+	}
+
+	return &searchResultsCommon{repos: repos}, matched
 }
 
 // reposToAdd determines which repositories should be included in the result set based on whether they fit in the subset
```

**Comment by stefanhengl** [comment]
interesting that the performances of all three approaches are statistically the same.

**Comment by keegancsmith** [comment]
Here is the full benchstat output with 5 runs each:

```
name                   old time/op  new time/op  delta
SearchRepositories      128ms ± 2%   135ms ± 2%   +5.18%  (p=0.008 n=5+5)
SearchRepositories-2   72.1ms ± 1%  69.3ms ± 1%   -3.80%  (p=0.016 n=4+5)
SearchRepositories-3   58.8ms ±10%  51.3ms ± 9%     ~     (p=0.056 n=5+5)
SearchRepositories-4   50.7ms ± 5%  48.2ms ±17%     ~     (p=0.841 n=5+5)
SearchRepositories-5   50.4ms ± 2%  43.9ms ± 7%  -12.90%  (p=0.016 n=4+5)
SearchRepositories-6   49.7ms ± 3%  42.2ms ±14%  -15.06%  (p=0.008 n=5+5)
SearchRepositories-7   48.3ms ± 3%  40.8ms ± 6%  -15.53%  (p=0.008 n=5+5)
SearchRepositories-8   47.8ms ± 2%  41.3ms ±12%  -13.60%  (p=0.008 n=5+5)
SearchRepositories-9   51.1ms ± 1%  40.6ms ± 5%  -20.53%  (p=0.008 n=5+5)
SearchRepositories-10  51.7ms ± 1%  40.6ms ± 2%  -21.42%  (p=0.008 n=5+5)
```

**Pull request merged**